### ISSUE TITLE
Problems with new Proxy-Vole 1.0.4

### ISSUE THREAD

**Issue by gschnepp** [MarkusBernhardt/proxy-vole#33]
I have problems with new Proxy-Vole 1.0.4 on different client computers. A rollback to 1.0.3 actually has helped in first situation, so it must have to do with the new proxy autodetection code. Problem is, I don't have direct access to these machines in different company networks of my customers.

**1. NoClassDefFoundError**
```
Exception in thread "main" java.lang.NoClassDefFoundError: Could not initialize class org.mozilla.javascript.VMBridge
        at org.mozilla.javascript.Context.exit(Context.java:434)
        at delight.rhinosandox.internal.RhinoSandboxImpl.assertContextFactory(RhinoSandboxImpl.java:68)
        at delight.rhinosandox.internal.RhinoSandboxImpl.evalWithGlobalScope(RhinoSandboxImpl.java:91)
        at com.github.markusbernhardt.proxy.selector.pac.JavaxPacScriptParser.setupEngine(JavaxPacScriptParser.java:80)
        at com.github.markusbernhardt.proxy.selector.pac.JavaxPacScriptParser.<init>(JavaxPacScriptParser.java:40)
        at com.github.markusbernhardt.proxy.selector.pac.PacProxySelector.selectEngine(PacProxySelector.java:76)
        at com.github.markusbernhardt.proxy.selector.pac.PacProxySelector.<init>(PacProxySelector.java:41)
        at com.github.markusbernhardt.proxy.util.ProxyUtil.buildPacSelectorForUrl(ProxyUtil.java:90)
        at com.github.markusbernhardt.proxy.search.browser.ie.IEProxySearchStrategy.createPacSelector(IEProxySearchStrategy.java:114)
        at com.github.markusbernhardt.proxy.search.browser.ie.IEProxySearchStrategy.getProxySelector(IEProxySearchStrategy.java:40)
        at com.github.markusbernhardt.proxy.ProxySearch.getProxySelector(ProxySearch.java:216)
        at de.credo.afplookup.license.RegistrationProperties.activateProperties(RegistrationProperties.java:122)
        at de.credo.afplookup.license.Registration.<clinit>(Registration.java:80)

```

**2. UnsatisfiedLinkError**
```
Exception in thread "main" java.lang.UnsatisfiedLinkError: [...]\jna-1224345557\jna7768348195287699756.dll: Can't find dependent libraries
        at java.lang.ClassLoader$NativeLibrary.load(Native Method)
        at java.lang.ClassLoader.loadLibrary0(Unknown Source)
        at java.lang.ClassLoader.loadLibrary(Unknown Source)
        at java.lang.Runtime.load0(Unknown Source)
        at java.lang.System.load(Unknown Source)
        at com.sun.jna.Native.loadNativeDispatchLibraryFromClasspath(Native.java:851)
        at com.sun.jna.Native.loadNativeDispatchLibrary(Native.java:826)
        at com.sun.jna.Native.<clinit>(Native.java:140)
        at com.sun.jna.Pointer.<clinit>(Pointer.java:41)
        at com.sun.jna.Structure.<clinit>(Structure.java:2078)
        at com.github.markusbernhardt.proxy.search.browser.ie.IEProxySearchStrategy.readIEProxyConfig(IEProxySearchStrategy.java:67)
        at com.github.markusbernhardt.proxy.search.browser.ie.IEProxySearchStrategy.getProxySelector(IEProxySearchStrategy.java:38)
        at com.github.markusbernhardt.proxy.ProxySearch.getProxySelector(ProxySearch.java:216)
        at de.credo.afplookup.license.RegistrationProperties.activateProperties(RegistrationProperties.java:122)
        at de.credo.afplookup.license.Registration.<clinit>(Registration.java:80)

```

I have had a look into the code in second case, and it seems code was not able to load WinHttp.dll. I don't know why, if this file doesn't exist on client computer or there's another problem in finding it (any app on this machine is deployed as [App-v package](https://de.wikipedia.org/wiki/Microsoft_Application_Virtualization), so software within this package may not have access to Windows system libraries by default).
Both cases are not related neccessarily. I'm putting them together in one issue anyway, because of the fact that these problems never raised the months before upgrading to Proxy-Vole 1.0.4.

**Any idea?** Both environments are complex company networks with internet access restricted very creatively, while both users where able to surf at the same time. As an example: After rollback to version 1.0.3 in first case license server access for my software wasn't able anyway. It turned out that user had also to fake user agent string of my software to _look like_ a browser. That's a functionality I had to implement months ago already for another customer, so faking user agent strings is not part of this proxy-vole 1.0.4 issue.

Regards
Guido

**Comment by gschnepp** [comment]
NoClassDefFound exception (bullet 1. above) has been reported as issue to delight-rhino-sandbox project on Github.

**Comment by gschnepp** [comment]
Library delight-rhino-sandbox has got a new release, 0.0.7. Project owner has no clue either, where this NoClassDefFoundError comes from, but wants to support with a free recompile: https://github.com/javadelight/delight-rhino-sandbox/issues/2#issuecomment-348120683

I've never used Git before and have no clue how to update POM by myself on Github. Is anyone able to update POM here to use this updated library with proxy-vole? Many thanks in advance!

**Comment by nerdydrew** [comment]
[Here's a branch with Rhino 0.0.7](https://github.com/nerdydrew/proxy-vole/tree/rhino_upgrade). WPAD worked for me on proxy-vole 1.0.4, so you should see if you can reproduce the NoClassDefFoundError with this.

**Comment by gschnepp** [comment]
Drew, thanks! Too busy here to publish it earlier, but today was the day. I'll report here if problems with new proxy-vole 1.0.5-SNAPSHOT in special environments will raise again - hopefully not. New Rhino (0.0.7) is much bigger in size than before, so chances are everything is included now.

**Comment by gschnepp** [comment]
I need you help, please!

No satisfaction with proxy-vole 1.0.5-SNAPSHOT, unfortunately!  :-(  I've just had another customer whose license registration crashed with same exception. Max from delight-rhino-sandbox has already been informed.

Looking at the code, it MUST have to do with invalid or unexpected PAC scripts, which were able to be processed up to proxy-vole 1.0.3 (without delight-rhino). I will try to get their PAC script from my customers, but: **Is anyone of you able to test with different PAC scripts from here?** Some classic ones, some with lists of proxies, some with nulls or otherwise irregular return values? This would really help!

Regards
Guido

**Comment by jstammi** [comment]
You are correct, this is the same problem.

Workarounds: 
 - You can install the visual c++ 2010 runtime on the machine
 - or you can exclude the proxy-vol-to-jna 4.2.2 dependency and add your own dependency to jna 4.5.1

**Comment by gschnepp** [comment]
Closing because
- bullet 2 will be fixed with proposed solution for #35, and
- bullet 1 had to do with illegally stripped down fat JAR (maven-shade-plugin "minimizeJar" has removed necessary dynamic classes from Rhino library, too)
